### PR TITLE
[Spark] Introduce V2 Checkpoint actions, table feature

### PR DESCRIPTION
**Pull request by prakharjain09**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Introduces actions for V2 Checkpoints - checkpointMetadata, sidecarFiles. Also introduces TableFeature for V2 Checkpoints.
Currently the feature is named with suffix `-under-development`. It will be removed in a later PR when all other pieces are merged.

Part of https://github.com/delta-io/delta/issues/1793 .

## How was this patch tested?

UTs

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala [hunk 1]
```diff
@@ -572,6 +572,35 @@ trait DeltaSQLConfBase {
...
+        """
+          |The file format to use for the top level checkpoint file in V2 Checkpoints.
+          | This can be set to either json or parquet. The appropriate format will be
+          | picked automatically if this config is not specified.
+          |""".stripMargin)
+      .stringConf
+      .checkValues(Set("json", "parquet"))
+      .createOptional
+
+  val EXPOSE_CHECKPOINT_V2_TABLE_FEATURE_FOR_TESTING =
```
Comment (ryan-johnson-databricks) [inline spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:592]: This is temporary, to be removed once the feature is ready for general use?

**Review by ryan-johnson-databricks — APPROVED** [review]
LGTM, just a few nits

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala [hunk 2]
```diff
@@ -684,3 +685,64 @@ object Checkpoints extends DeltaLogging {
...
+   * This needs [[V2CheckpointTableFeature]] to be enabled on the table.
+   */
+  case object V2 extends Policy("v2")
+
+  /** ALl checkpoint policies */
+  val ALL: Seq[Policy] = Seq(Classic, V2)
+
+  /** Converts a `name` String into a [[Policy]] */
+  def fromName(name: String): Policy = ALL.find(_.name == name).getOrElse {
+    throw new IllegalStateException(s"Invalid policy $name")
```
[the remark this answers is not in the document]
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:746]: IllegalArgument suits better.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala [hunk 3]
```diff
@@ -624,6 +624,15 @@ trait DeltaConfigsBase extends DeltaLogging {
     "must be Serializable"
   )
 
+  /** Policy to decide what kind of checkpoint to write to a table. */
+  val CHECKPOINT_POLICY = buildConfig[CheckpointPolicy.Policy](
+    key = "checkpointPolicy-dev",
+    defaultValue = CheckpointPolicy.Classic.name,
+    fromString = str => CheckpointPolicy.fromName(str),
+    validationFunction = (v => CheckpointPolicy.ALL.exists(_.name == v.name)),
```
[the remark this answers is not in the document]
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala:632]: I think most of the other table properties don't handle lower/upper case - following the same pattern.

**Comment by tdas** [comment]
LGTM.

**Comment by vkorukanti** [comment]
Merged as 22ade20b20e34b6ae18715d81288b26e0f53194e

**Pull request closed without merging**